File: 0209-Minimum-Size-Subarray-Sum/java-0209/src/Solution1.java
/// https://leetcode.com/problems/minimum-size-subarray-sum/description/
/// Author : liuyubobobo
/// Time   : 2017-11-13

// Sum Prefix
// Time Complexity: O(n^2)
// Space Complexity: O(n)
public class Solution1 {

    public int minSubArrayLen(int s, int[] nums) {

        if(s <= 0 || nums == null)
            throw new IllegalArgumentException("Illigal Arguments");

        int[] sums = new int[nums.length + 1];
        sums[0] = 0;
        for(int i = 1 ; i <= nums.length ; i ++)
            sums[i] = sums[i-1] + nums[i-1];

        int res = nums.length + 1;
        for(int l = 0 ; l < nums.length ; l ++)
            for(int r = l ; r < nums.length ; r ++)
                if(sums[r+1] - sums[l] >= s)
                    res = Math.min(res, r - l + 1);

        if(res == nums.length + 1)
            return 0;

        return res;
    }

    public static void main(String[] args) {

        int[] nums = {2, 3, 1, 2, 4, 3};
        int s = 7;
        System.out.println((new Solution1()).minSubArrayLen(s, nums));
    }
}

File: 0152-Maximum-Product-Subarray/py-0152/Solution1.py
# Source : https://leetcode.com/problems/maximum-product-subarray/
# Author : penpenps
# Time   : 2019-07-21

from typing import List

# Travesal nums, use maxVal and minVal to keep max and min product end with current element
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        minVal, maxVal = nums[0], nums[0]
        ans = nums[0]
        for i in range(1, n):
            # If negative number, it should swap min and max for sign switching
            if nums[i] < 0:
                minVal, maxVal = maxVal, minVal
            maxVal = max(nums[i]*maxVal, nums[i])
            minVal = min(nums[i]*minVal, nums[i])
            ans = max(maxVal, ans)

        return ans
        

if __name__ == '__main__':
	s = Solution()
	nums = [2,3,-2,4]
	answer = s.maxProduct(nums)
	print(answer)
